### DISCUSSION TITLE
CPU High 🔥

### DISCUSSION THREAD

**Discussion by FeMaffezzolli** [omar-dulaimi/prisma-trpc-generator#60] [Q&A]
I just followed the setup steps successfully and then my CPU usage (VSCode Code Help, which does the Intellisense) is going beyond 100% which basically breaks the VSCode usability and typescript.

I have 10 models in my Prisma schema (not much I think). Besides I have a 2021 Macbook Pro M1 Pro 16GB ram.
So I suppo I'm missing something.

![image](https://user-images.githubusercontent.com/26548010/211126619-d87dab85-15a8-41f2-9572-598231105731.png)

```prisma
generator trpc {
  provider       = "prisma-trpc-generator"
  withMiddleware = false
  withShield     = false
  contextPath    = "../src/server/trpc/context"
  // trpcOptionsPath = "../src/server/trpc/trpcOptions"
  output         = "../src/server/trpc/generated"
}
```
Can someone help?

**Comment by cqh963852** [comment]
Meeting the same problem. (I have only 5 models.)

I think it is caused by typescript language server. 

Because when I type, the typescript prompt is very slow.

**Comment by omar-dulaimi** [comment]
It's something Typescript cause when there are many files. Other generators for Prisma have suffered of this issue before and still.  I'm thinking of providing an option to export all routers and schemas in a single file(or two). That should fix the problem I guess.

Would that be useful to you?

**Reply by cqh963852** [reply]
I've come across situations where the code is in one file. 

we use graphql codegen generate our component about 100 component.  each component has 1000 line.

It take me so many memory, even crashes eslint.

Maybe should tell typescript to ignore the generated files.

**Reply by maxime4000** [reply]
I think the option between single file and multi files would be nice. Also subfolder when multi File (in schemas folder only)

**Comment by FeMaffezzolli** [comment]
> It's something Typescript cause when there are many files. Other generators for Prisma have suffered of this issue before and still.  I'm thinking of providing an option to export all routers and schemas in a single file(or two). That should fix the problem I guess.
> 
> 
> 
> Would that be useful to you?

I don't know typescript in depth enough to know how it works under the hood so I can't say if using one file versus multiple will fix the problem. Do you think so?

Maybe we should get some inspiration from Prisma itself. For example, why does it create the type file in the node_modules?

**Reply by arnm** [reply]
I tried another zod prisma generator library and had the same issue with just one output file with all the schemas vs schemas in their own files :(

**Comment by maxime4000** [comment]
I don't think this is doable. The issue come from Zod itself. Zod infer a lot to type object. Those type object created by Zod are dynamically made with the schema. If you are writting Zod schemas, it won't affect you much because you are only launching the LSP for that definition your working on. Because it's hundred or thousand of schemas you generate in one go with this generator. You overload the Typescript's LSP and it try to type every schemas at the same time. 

In fact, I would expect that if the generator wouldn't delete the file just before generating it again, it would be less CPU heavy. Typescript wouldn't have to analyse every definition, because he would know that this definition hasn't changed from the last time he analyse it. 

Tbh, I don't know if this would help. File would still be overwritten which would trigger Typescript to analyse the file. 

My solution is to comment the generator when I don't want to update Zod/tRPC and uncomment when I need it.

**Comment by arnm** [comment]
I just asked the same question in #69 this is really killing the DX :(

**Comment by omar-dulaimi** [comment]
Same root cause as #69, so I have written the full answer there and will keep the two together rather than split it.

The short version, and it corrects what I told you at the time: I said this was TypeScript struggling with a large number of files, and proposed bundling everything into one or two files. **@maxime4000 was right that this is Zod's inference, not the file count.** Each generated schema produces a large structural type and `z.infer` over deep Prisma input types expands into work the language server repeats constantly. Fewer, larger files would not have fixed it. I never shipped that option, which turns out to be the right call for the wrong reason.

What actually reduces it: generate fewer operations. Operation-level exclusion landed in #68, and every operation you skip removes a whole family of input types from the checker's graph. Ten models should not melt an M1 Pro, and the reason it did is that the generator emitted every operation for every model by default.

See #69 for the full write-up and the ordered list of what helps.
